### ISSUE TITLE
Assertion failure: !(pollset [i].revents & POLLNVAL) (poll.cpp:157)

### ISSUE THREAD

**Issue by sigiesec** [zeromq/libzmq#2895]
# Issue description

An assertion failure occurs in the following line:
https://github.com/zeromq/libzmq/blob/3226b8ebddd9c6c738ba42986822c26418a49afb/src/poll.cpp#L157

The fd member of the offending entry in pollset in that case has the value -1 / retired_fd.

Apparently, this was already the case when poll was called, which caused poll to set the revents of the entry to POLLNVAL.

I found that it is possible to call add_fd with a fd_ parameter of retired_fd/-1, which is not checked, so I assume this has happened somehow, but I do not know via which call chain this happened.

# Environment

* libzmq version (commit hash if unreleased): 4.2.3 built with ZMQ_USE_POLL
* OS: Windows

# Minimal test code / Steps to reproduce the issue

Unfortunately, I have not been able to reproduce this issue. It occurs if some networking issues happen, such as firewall configurations between peers change, or the local network settings are reconfigured. I don't know yet what exactly triggers the issue, but it has occurred several times on different machines.

# What's the actual result? (include assertion message & call stack if applicable)

Assertion failure: !(pollset [i].revents & POLLNVAL) within a zmq::poll_t::worker_routine thread.

# What's the expected result?

No assertion failure, operation continues.

**Comment by sigiesec** [comment]
I investigated where an INVALID_SOCKET might come from and found the following:
* zmq::signaler_t::signaler_t calls make_fdpair, which might fail: https://github.com/zeromq/libzmq/blob/4518e0cc35ffae84b077044c071b3c78e5d4237c/src/signaler.cpp#L127 How is this case handled? Since this is a constructor, it has no return value. If this is not handled properly, zmq::signaler_t::get_fd will return INVALID_SOCKET afterwards, which is not checked at least in some places, e.g. in zmq::io_thread_t::io_thread_t: https://github.com/zeromq/libzmq/blob/4518e0cc35ffae84b077044c071b3c78e5d4237c/src/io_thread.cpp#L45
This seems related to #1558, which was fixed by #1564, which changes a location similar to the above-mentioned io_thread_t constructor. I think that should be changed in the same way. @c-rack Since you contributed #1564, what do you think?

**Comment by sigiesec** [comment]
A similar situation is here: https://github.com/zeromq/libzmq/blob/4518e0cc35ffae84b077044c071b3c78e5d4237c/src/socket_base.cpp#L1330
Maybe fd can never be retired_fd/-1 here for some reason, but then an assertion should be added here.

**Comment by sigiesec** [comment]
I tried to create a large number of context, and one socket in each context. After about 4000 contexts/sockets, zmq_socket fails with an error code of ENOBUFS, which is out of the specified behaviour of zmq_socket. However, probably the same error might occur during make_fdpair. This is most probably caused by the global limit of ephemeral ports being exceeded (https://blogs.technet.microsoft.com/tristank/2008/03/11/maxuserport-what-it-is-what-it-does-when-its-important/). The appropriate reaction would be to retry after some time IMO.

**Comment by sigiesec** [comment]
With the following code I was able to trigger the problem zmq::signaler_t:
```
    std::vector<void *> contexts;
    std::vector<void *> sockets;

    const size_t count = 10000;

    for (size_t i = 0; i < count; ++i) {
        fprintf (stderr, ".");
        contexts.emplace_back (zmq_ctx_new ());

        if (contexts.back () == nullptr) {
            fprintf (
              stderr,
              "\ncontext creation failed after %i contexts, errno = %i\n",
              (int) i, errno);
            break;
        }

        void *bind_socket = zmq_socket (contexts.back (), ZMQ_DEALER);
        void *connect_socket = zmq_socket (contexts.back (), ZMQ_DEALER);
        if (bind_socket == nullptr || connect_socket == nullptr) {
            if (bind_socket != nullptr)
                zmq_close (bind_socket);
            fprintf (stderr,
                     "\nsocket creation failed after %i contexts, errno = %i\n",
                     (int) i, errno);
            continue;
        }

        int rc = zmq_bind (bind_socket, "inproc://test");
        if (rc != 0) {
            fprintf (stderr,
                     "\nsocket bind failed after %i contexts, errno = %i\n",
                     (int) i, errno);
            zmq_close (bind_socket);
            zmq_close (connect_socket);
            continue;
        }

        rc = zmq_connect (connect_socket, "inproc://test");
        if (rc != 0) {
            fprintf (stderr,
                     "\nsocket connect failed after %i contexts, errno = %i\n",
                     (int) i, errno);
            zmq_close (bind_socket);
            zmq_close (connect_socket);
            continue;
        }

        sockets.emplace_back (bind_socket);
        sockets.emplace_back (connect_socket);
    }
```

I then get an assertion failure here:
```
Assertion failed: Socket operation on non-socket (D:\Dev\libzmq\src\signaler.cpp:192)
```
i.e. here:
https://github.com/zeromq/libzmq/blob/4518e0cc35ffae84b077044c071b3c78e5d4237c/src/signaler.cpp#L192

**Comment by sigiesec** [comment]
Not completely fixed, as seen e.g. in https://ci.appveyor.com/project/sigiesec/libzmq/build/build-378/job/0hkiqeywv9op2a9w#L1183

**Comment by Crypto2** [comment]
I'm also having this same problem with 4.2.3 with ZMQ_USE_POLL

**Comment by bluca** [comment]
It's fixed in 4.2.5

**Comment by sigiesec** [comment]
@Crypto2 Note however that poller=poll is broken on Windows, and will no longer be offered as an option in master. Either use poller=select with 4.2.5, or poller=epoll with master, support for which was only recently added under Windows.
